21.md

```diff
@@ -0,0 +1,202 @@
+
+The extension is opt-in and backwards compatible. Further, a supporting `WALLET` can always tell if a `SERVICE` is also supporting beforehand so the communication is never ambiguous.
+
+### Wallet-side first request
+
+The first request is unchanged from the base specification.
+
+### Service-side first response
+
+`SERVICE` must alter its JSON response to the first callback to include a `currencies` field, as follows:
+
+```typescript
+type BaseResponse = {
+    tag: "payRequest",
+    metadata: string,
+    callback: string,
+    maxSendable: number,
+    minSendable: number
+};
+
+type Currency = {
+  name: string, // Name of the currency. E.g.: Reais
+  symbol: string, // Symbol of the currency. E.g.: R$
+  decimals: number, // Number of decimal places. E.g.: 2
+  price: number, // Number of currency per BTC. E.g.: 185000
+  convertable?: bool // Whether the currency can be converted into.
+}
+
+type ExtendedResponse = BaseResponse & {
+  currencies: { [key: string]: Currency }
+}
+```
+
+```diff
+{
+  "tag": "payRequest",
+  "metadata": '[["text/plain","$kenu ⚡ bipa.app"]]',
+  "callback": "https://api.bipa.app/ln/request/invoice/kenu",
+  "maxSendable": 1000000000,
+  "minSendable": 1000,
++ "currencies": {
++   "BRL": {
++     "name": "Real",
++     "symbol": "R$",
++     "decimals": 2,
++     "price": 185000,
++     "convertable": true
++   }
++ }
+}
+```
+
+- The inclusion of the `currencies` field implies the support of this extension
+- The inclusion of a `currency` implies it can be used for denomination of an amount
+- The inclusion of a `convertable currency` implies the `SERVICE` can quote and guarantee a price for given currency
+- The price returned on this response is not guaranteed by the `SERVICE` and is subject to change
+- The `key` of a `currency` will be used as an identifier for the next request
+- The `key` and other information must be according to [ISO-4217](https://en.wikipedia.org/wiki/ISO_4217) for currencies that are included in it
+
+### Wallet-side second request
+
+Upon seeing the `currencies` field on the response from `SERVICE`, a `WALLET` may show the user it has the option of denominating the amount in different currencies or even of sending sats to be credited as the different currency for the receiver.
+
+After gathering the input from the user, it must be encoded into the callback as query parameters as follows.
+
+The most general case, when the user is denominating the amount in `CURRENCY1`, but wants it to be converted to `CURRENCY2` on `SERVICE`.
```

Comment:
I'm slightly confused on who's responsible for knowing the conversions here. Is it always the payee service that needs to know both CURRENCY1 and CURRENCY2 and their exchange rates to/from bitcoin?

This differs slightly from the UMA flow (which may be fine since it's a slightly different use case). With UMA , the sender's service (wallet/custodian/exchange/etc) would be responsible for converting from the sending currency to bitcoin before traveling over lightning. For that reason, its communication of its fees and conversion rate to the user are outside of the scope of UMA/LNURL and just between it and its user. It's really just always bitcoin being sent to the receiver and the payee service doesn't need to care what currency it was before it was bitcoin.

Reply:
You understood it correctly, but let me try to clarify. If I ask for a lightning invoice of 1BRL, the payee service will have know the conversion rate and return me with an invoice of about 538 sats. Then if I ask the payment to covnerted into BRL, then the payee service (again) will be responsible for performing the exchange and crediting the payee with BRL.

So the payer never has to know any exchange information, it's all on the payee side. It might want to check exchange rate or something, but it's optional and out of scope of the proposal.

About your second point, I'm actually really glad you mentioned it, because what I call colloquially "source exchange data" was part of the [first draft](https://github.com/bipa-app/lnurl-rfc/pull/1) I thought about this subject. In order to really "simulate" a remittance, we need the payee service to be able to tell which currency the BTC came from and at what cost.

I decided not to add this part to this spec because I think it can be thought of a separate complementary LUD, which as soon as this one is done I'll work on immediately. So at this point, the 'source exchange data' is outside the scope of this proposal and LNURL, like you said, and bitcoin is the only thing being sent and the payee service doesn't know anything about the payer sourcing of bitcoin.

Does that clear things out? I'm happy to keep discussing if things are not clear

Reply:
Nice, I'm glad to hear you've already thought a bit about the source exchange data! Having it as separate to this proposal sounds fine to me as long as it's extensible enough to add that ability in the future.

> In order to really "simulate" a remittance, we need the payee service to be able to tell which currency the BTC came from and at what cost.

Do you mind explaining why the payee service needs to know this? I don't have a ton of knowledge about traditional remittences, but I'd love to learn :-).

Reply:
It's actually so the user can have a nice experience typing (meaning, I can preview how much will it receive or spend while the typing is happening) without actually using the callback (which can be expensive in some implementations). :D